pkg/volume/rbd/rbd.go

```diff
@@ -409,15 +443,187 @@ type rbd struct {
 	mounter  *mount.SafeFormatAndMount
 	exec     mount.Exec
 	// Utility interface that provides API calls to the provider to attach/detach disks.
-	manager diskManager
+	manager rbdUtil
 	volume.MetricsProvider
 }
 
+var _ volume.Volume = &rbd{}
+
 func (rbd *rbd) GetPath() string {
 	// safe to use PodVolumeDir now: volume teardown occurs before pod is cleaned up
 	return getPath(rbd.podUID, rbd.volName, rbd.plugin.host)
 }
 
+// newRBD creates a new rbd.
+func newRBD(podUID types.UID, volName string, image string, pool string, readOnly bool, plugin *rbdPlugin, manager rbdUtil) *rbd {
+	return &rbd{
+		podUID:          podUID,
+		volName:         volName,
+		Image:           image,
+		Pool:            pool,
+		ReadOnly:        readOnly,
+		plugin:          plugin,
+		mounter:         plugin.mounter,
+		exec:            plugin.exec,
+		manager:         manager,
+		MetricsProvider: volume.NewMetricsStatFS(getPath(podUID, volName, plugin.host)),
+	}
+}
+
+// rbdAttacher implements volume.Attacher interface.
+type rbdAttacher struct {
+	plugin  *rbdPlugin
+	manager rbdUtil
+}
+
+var _ volume.Attacher = &rbdAttacher{}
+
+// Attach implements Attacher.Attach method. It requires a exclusive lock of
+// image for given node from ceph cluster.
+// It's called by kube-controller-manager, if kubelet's --enable-controller-attach-detach flag is true.
+// Otherwise, it's called by kubelet itself.
+func (attacher *rbdAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string, error) {
+	glog.V(4).Infof("rbd: attaching %s onto %s", spec.Name(), nodeName)
+	mounter, err := attacher.plugin.createMounterFromVolumeSpec(spec)
+	if err != nil {
+		return "", err
+	}
+	err = attacher.manager.Fencing(*mounter, string(nodeName))
```

Comment:
cannot fence here: containerized controller-master doesn't have rbd cli in it. Check gcr.io/google_containers/kube-controller-manager-amd64